### DISCUSSION TITLE
O que é o conceito de "closure" em JavaScript e por que ele é importante?

### DISCUSSION THREAD

**Discussion by oeduardostein** [onicolasdelfino/onicolasdelfino#13] [Q&A] [answered]
O que é o conceito de "closure" em JavaScript e por que ele é importante?

**Comment by onicolasdelfino** [comment, marked as answer]
Um closure em JavaScript ocorre quando uma função "lembra" do escopo em que foi criada, mesmo após esse escopo ter sido encerrado. Isso permite que uma função interna acesse variáveis de uma função externa mesmo após a execução da função externa ter terminado.

Aqui está um exemplo simples:

javascript
Copiar código
function criarContador() {
    let contador = 0;
    return function() {
        contador++;
        console.log(contador);
    };
}

const incrementar = criarContador();
incrementar();  // Saída: 1
incrementar();  // Saída: 2
Neste exemplo, a função interna "lembra" da variável contador da função externa criarContador, mesmo após criarContador já ter retornado. Esse comportamento é chamado de closure.

Closures são importantes porque permitem que você crie funções com escopos "privados", mantendo variáveis acessíveis apenas dentro do contexto da função interna. Isso ajuda a preservar o estado entre chamadas de função e a evitar conflitos no escopo global.
